provision-vm.yml

```
@@ -0,0 +1,61 @@
+- name: Install To-Do App on new web servers
+  hosts: webservers
+  remote_user: ec2-user
+  vars_prompt:
+    - name: trello_api_key
+      prompt: Please enter a Trello API key.
+    - name: trello_api_token
+      prompt: Please enter a Trello API token.
+  tasks:
+  - name: Install Git
+    ansible.builtin.yum:
+      name: git
+      state: installed
+    become: yes
+
+  - name: Install Python3
+    ansible.builtin.yum:
+      name: python3
+      state: latest
+    become: yes
+
+  - name: Install Poetry
+    ansible.builtin.shell:
+      cmd: curl -sSL https://install.python-poetry.org | python3 -
+      creates: ~/.local/bin/poetry
+
+  - name: Create directory for the To-Do App code
+    ansible.builtin.file:
+      path: /opt/todoapp
+      state: directory
+      owner: ec2-user
+    become: yes
+
+  - name: Get the latest version of the code with Git
+    ansible.builtin.git:
+      repo: 'https://github.com/oscar-richardson-softwire/DevOps-Course-Starter.git'
+      dest: /opt/todoapp
+      version: main
+
+  - name: Install project dependencies
+    ansible.builtin.shell:
+      cmd: cd /opt/todoapp; /home/ec2-user/.local/bin/poetry install
+      creates: /opt/todoapp/.venv
```

Two things to consider here:
1. Very minor, nut worth noting there's a `chdir` (change directory) option for shell commands, so that we can avoid chaining the cd for a slightly nicer syntax
2. Not one I've tested, but worth considering whether there's a risk of using `creates` here - _do_ we want to skip the installation step if a virtual environment already exists? E.g. if we've updated the dependency list in `pyproject.toml`, this could risk skipping the step of installing that new dependency
  i. That might depend on whether or not the git command removes the `.venv` folder normally, but I _suspect_ it doesn't

```suggestion
  - name: Install project dependencies
    ansible.builtin.shell:
      cmd: /home/ec2-user/.local/bin/poetry install
      creates: /opt/todoapp/.venv
      chdir: /opt/todoapp/
```

---

Thanks for this, I agree that the `chdir` syntax is nicer and have applied this change 😄 Also, good spot RE `creates` - I've tested it and the git command _doesn't_ remove the `.venv` folder, so I've removed the `creates` line to ensure that the dependencies are reinstalled.